tari-project/tari
# [wallet-ffi] Update `wallet_create` to take seed words for recovery

## Description


- FFI BREAKING CHANGE: FFI function `wallet_create` now takes an optional seed words arg, if provided this sets the wallet master recovery key from which the comms public key and spending keys are derived from. 

- **Pre-alpha** quality NodeJS FFI wallet client in `applications/ffi_client` that can create a new wallet or recover an old wallet via the FFI lib. Janky and unreliable as hell 🔪

## Motivation and Context


Allows wallet recovery from FFI which is required for the mobile clients. The ffi_client can be used to test recovery via FFI. 

## How Has This Been Tested?



1. New Rust test to check that a wallet created from seed words has matching seed words and public key. 
1. Recovered a wallet via FFI with the following steps: 
- Build the FFI lib: `cargo build -p tari_wallet_ffi --release --lib`
- Copy the lib into the ffi_client folder: `cp target/release/libtari_wallet_ffi.dylib /path/to/ffi_client/`
- Run `npm install` in the ffi_client folder
- Set your seed words as env var `export SEED_WORDS="two small frozen ..."`
- Run `npm run recovery`, if error or crash run again until success
- Recovered wallet is in `ffi_client/recovery`  folder 

- Running FFI recovery (multiple runs, some failures)
```
> node recovery.js

Set up library...
Create Tor transport...
Create Comms config...
Create Wallet from seed words...
Set base node peer... 0c3fe3c23866ed3827e1cd72aae0c9d364d860d597993104e90d9a9401e52f05
Starting recovery...
recovery:  { current: 0 } { total: 1 }
recovery:  { current: 12679 } { total: 14093 }
recovery:  { current: 12688 } { total: 14093 }
===
Balance
===
  available :  0
  pending_in:  0
===
^CexitHandler
options { exit: true, signal: 'SIGINT' }
exitCode SIGINT
-----

> node recovery.js

Set up library...
Create Tor transport...
Create Comms config...
Create Wallet from seed words...
===
Balance
===
  available :  1249005213456
  pending_in:  0
===
Set base node peer... 0c3fe3c23866ed3827e1cd72aae0c9d364d860d597993104e90d9a9401e52f05
Starting recovery...
started true 0
recovery:  { current: 0 } { total: 1 }
===
Balance
===
  available :  1249005213456
  pending_in:  0
===
recovery:  { current: 14093 } { total: 14095 }
===
Balance
===
  available :  1249005213456
  pending_in:  0
===
recovery:  { current: 14095 } { total: 14095 }
===
Balance
===
  available :  1249005213456
  pending_in:  0
===
exitHandler
options { cleanup: true, signal: 'exit' }
exitCode 0
===
Balance
===
  available :  1249005213456
  pending_in:  0
===
Exiting...

```

- FFI recovered wallet opened in console wallet: 
![image](https://user-images.githubusercontent.com/351403/120636983-a0505d00-c46e-11eb-9c8b-fec739afbb61.png)


## Checklist:

* [x] I'm merging against the `tari-script` branch.
* [x] I have squashed my commits into a single commit.
